eth/downloader/queue.go

```diff
@@ -363,28 +383,28 @@ func (q *queue) WaitResults() []*fetchResult {
 		}
 		// Advance the expected block number of the first cache entry.
 		q.resultOffset += uint64(nproc)
+
+		// Recalculate the result item weights to prevent memory exhaustion
+		for _, result := range results {
+			var size common.StorageSize
+
+			blob, _ := rlp.EncodeToBytes(result.Receipts) // TODO(karalabe): Whoah, nice hack, super optimal
```

Comment:
:) 
Don't forget that TODO. 
Q: Wouldn't it be pretty simple to implement `rlp.CountBytes(...)`, which recursively just counts but does not actually allocate memory?

Reply:
Actually, you can do this instead, in rlp.go:
```
// EncodedBytesize returns the size that RLP encoding of val would have,
// if EncodeToBytes was called
func EncodedBytesize(val interface{}) (int, error) {
	eb := encbufPool.Get().(*encbuf)
	defer encbufPool.Put(eb)
	eb.reset()
	if err := eb.encode(val); err != nil {
		return 0, err
	}
	return eb.size(), nil
}
```
That'll roughly halve the memory allocations there.